Godeps: bump appc/spec to v0.6.0

iaguis opened the pull request:
This is an attempt to bump the spec to v0.6.0. I ran into some problems:

### Docker2aci and goaci need to be bumped too

No problems with this.

### Entrypoint annotations

Since characters allowed in ACNames are more restricted now, the existing annotations don't work because they have slashes. E.g. `coreos.com/rkt/stage1/run`. I changed them to things like `stage1-run` for now but we should change annotation's type to ACIdentifier.

### We were using ImageManifest.Name and RuntimeApp.Name interchangeably

They now have a different type, ImageManifest.Name has type ACIdentifier and RuntimeApp.Name ACName.

This is easy to solve sometimes (just reference ImageName instead of App) but sometimes it's not so clear. For example, when generating a Pod manifest the user provides an Image Name but what ends up in the Pod manifest is an RuntimeApp.Name (there's even a place in the code where we force App names and Image names to be the same!). We need some sort of translation.

What I implemented is a translation where I split the Image Name with `/` as separator, take the last element and then use SanitizeACName. For example, `coreos.com/cool-apps/etcd.production` is translated to `etcd-production`. Suggestions welcome. Related: #336

In other places we use the Image Name to search for an App, I implemented a function to search Apps by Image Name. This function should probably be moved to the spec but for now it lives in rkt/common.

Fixes #1032

yifan-gu commented on stage1/init/pod.go:
```diff
@@ -346,7 +346,7 @@ func (p *Pod) writeEnvFile(env types.Environment, id types.Hash) error {
 func (p *Pod) PodToSystemd(interactive bool) error {
 	for _, am := range p.Apps {
-		ra := p.Manifest.Apps.Get(am.Name)
+		ra := common.GetAppByImageName(p.Manifest.Apps, am.Name)
```
I think we can just iterate over p.Manifest.Apps now? As we can guarantee p.Manifest.Apps[i].App is non-nil after #1002

yifan-gu commented on stage1/init/pod.go:
```diff
@@ -443,7 +443,7 @@ func (p *Pod) PodToNspawnArgs() ([]string, error) {
 
 	for _, am := range p.Apps {
-		ra := p.Manifest.Apps.Get(am.Name)
+		ra := common.GetAppByImageName(p.Manifest.Apps, am.Name)
```
ditto above

iaguis commented:
So I took @yifan-gu's branch from #1038, renamed entrypoints (although I still think the entrypoint type should namespeced and hence an ACIdentifier) and updated the docs like @alban suggested in [#1038 (comment)](https://github.com/coreos/rkt/pull/1038#discussion_r32403100). However, as he says in the same comment, changing entrypoints' values can be problematic:

```
It is an ABI break between stage0 and stage1.
So rkt gc will fail to garbage collect old pods from a previous version. Is it ok?
```

Is renaming entrypoints acceptable or should we wait for a future appc/spec v0.6.1 with ACIdentifier entrypoints and keep the old names?

What about the translation of ImageManifest.Name to App.Name? Is assuming there won't be apps with the same name in a different namespace (e.g. `coreos.com/etcd` and `example.org/etcd`) in a pod reasonable for now?

yifan-gu commented:
> What about the translation of ImageManifest.Name to App.Name? Is assuming there won't be apps with the same name in a different namespace (e.g. coreos.com/etcd and example.org/etcd) in a pod reasonable for now?

I think it's OK for now, until use cases like running same images in a pod appears, then we can revisit this, maybe add some random suffix?

jonboulle commented:
Per OOB discussion dropping in favour of #1038
